### ISSUE TITLE
should we have an extension point for model transforms out of tree?

### ISSUE THREAD

**Issue by vkuzo** [pytorch/torchtitan#790]
In [torchao](https://github.com/pytorch/ao), we have various low precision training features which are in prototype: MX, int8, bitnet.  While we expect most of these to eventually end up in the main torchao APIs, it often takes ~months for a prototype to graduate.

torchtitan is extremely useful for helping us test low precision prototypes in real-world settings.  For now, we've been creating unlanded PRs to test functionality (examples: https://github.com/pytorch/torchtitan/pull/614, https://github.com/pytorch/torchtitan/pull/778).  Would torchtitan consider building an extension point to support this kind of experimentation fully out-of-tree?

An example of how this could look like:
1. torchtitan provides a "model transformation" hook that it calls at a specified point in the initialization stage (for quantization, that should be after model init and before parallelization / torch.compile)
2. user can provide a custom pass to transform the model (such as a prototype low precision training conversion pass)

I'm not entirely sure on how this hook would be implemented since the current interface of torchtitan is CLI based, but wanted to share the request and start the discussion.

**Comment by vkuzo** [comment]
cc @awgu , @tianyu-l , @weifengpy

**Comment by tianyu-l** [comment]
more than happy to work together on this!

Before we explore solutions, may I ask questions to better understand your requests
- what would you like to achieve ideally, e.g. do you want to use torchtitan as a library without cloning it? or more specifically demonstrate your work in torchao by importing torchtitan as a library?
- (maybe repeating the question above) what are the "pain points" you are experiencing in the current ways (PR / branch / fork)

**Comment by tianyu-l** [comment]
@balancap 
Thank you very much for the suggestions! Generalizing the Float8Handler sounds a reasonable thing to do. 

I have a question though:
I'm assuming `convert` would be applied before the parallelization & compilation code. It sounds viable if the underlying change in `convert` stay compatible with those code. E.g. one might want to modify a (llama) model in a "parallelization-breaking" way, or register a completely new model (see https://github.com/pytorch/torchtitan/issues/282). In such cases, do you think it's better to only support the forking way, not the library/submodule way?

another general question:
Could you share that in your case, what parts of torchtitan you'd keep and what parts you'd modify?
E.g. keeping the model, data loading, parallelization; modifying the dataset, optimizer / lr scheduler, FlexAttention / kernels

cc: @vkuzo

**Comment by balancap** [comment]
Indeed, I would keep `convert` before the parallelization & compilation code (like `Float8Handler` https://github.com/pytorch/torchtitan/blob/main/train.py#L116), and the user should be aware to that they need to stay compatible with those passes. Most quantization methods I can think of should satisfy that, as wel as swapping simple layers. If some people want to do deeper modifications, I agree forking is the way (and I tend to encourage it in our team if I feel a project needs to deeply modify TT). It should probably be very clear and explicit in `ModelHandler` definition what is the intended usecase, and what is beyond that scope.

For the kind of projects we have in mind, it is in line with what you're saying:
**Keep:** Parallelization, training loop, checkpointing, Llama model; => robust, but simple enough base infra with good MFU .
**Change:** Some layers in model (quantized linear, flex attention, liger kernels, ...), optimizer, LR scheduler and dataset. => Things which are fairly safe to modify without breaking parallelism or collapsing MFU.

**Comment by vkuzo** [comment]
> what would you like to achieve ideally, e.g. do you want to use torchtitan as a library without cloning it? or more specifically demonstrate your work in torchao by importing torchtitan as a library?

yes, I think that's a good way to describe it!

> Keep: Parallelization, training loop, checkpointing, Llama model; => robust, but simple enough base infra with good MFU .
> Change: Some layers in model (quantized linear, flex attention, liger kernels, ...), optimizer, LR scheduler and dataset. => Things which are fairly safe to modify without breaking parallelism or collapsing MFU.

yes, that matches 90% of my use cases as well.  For the other 10% forking is fine as well, as I have been doing!

**Comment by fegin** [comment]
@balancap I'm curious why do`pre_optimizer_hook` and `post_opotimizer_hook` mean? Do the the pre_hook and post_hook of optimizer.step()? If so, shouldn't these hooks belong to `Optimizer` class? Or `ModelHandler` is more close to `TrainLoop` class?

**Comment by tianyu-l** [comment]
@fegin I think the `float8_handler` in train.py is a good example. For other use cases, @balancap please add.
In particular, are there motivations other than quantization to have the `pre/post_optimizer_hooks`?

> Keep: Parallelization, training loop, checkpointing, Llama model; => robust, but simple enough base infra with good MFU .
Change: Some layers in model (quantized linear, flex attention, liger kernels, ...), optimizer, LR scheduler and dataset. => Things which are fairly safe to modify without breaking parallelism or collapsing MFU.

Besides, I'd like to understand your needs more precisely:
In order change optimizer / lr scheduler, I believe today one has to slightly modify the train loop. (Do you want that part also be configurable from config?) Also if installed as a library, one still needs to copy the train loop anyway?
And if train loop needs to be modified, you can still write your own handler out-of-TT, but just can't easily config things in toml.

Could you clarify a bit?

**Comment by balancap** [comment]
We are using TorchTitan as a git submodule, so our training script look like that:

**Comment by balancap** [comment]
We are using TorchTitan as a git submodule, so our training script look like that:
```python
import torch
import wandb

# First import for MonkeyPatching TorchTitan.
import my_llm_training_module  # noqa: F401
from my_llm_training_module import JobConfig # Addiitional Job config options.

from torchtitan import train as tt_train
from torchtitan.logging import logger


def main():
    config = JobConfig()
    config.parse_args()
    # Setup wandb...
    # Main TorchTitan training setup & loop
    try:
        tt_train.main(config)
        torch.distributed.destroy_process_group()
    except Exception as e:
        # Error logging before process ends, to record in W&B.
        # Keeping formatting similar to `torchrun` error output.
        logger.error("--- Logging error --")
        logger.error(f"{type(e).__module__}.{type(e).__qualname__}: {e}", exc_info=True)
        raise
    finally:
        # Note keeping W&B init + finish in `main` for clean exception handling.
        wandb.finish()

if __name__ == "__main__":
    main()
```
So for some projects, it has the benefit of avoiding copy-pasting the train loop (which means potentially complex testing to make sure we don't break things). Definitely not a fit all approach, but it has some benefits.

At the moment, we are changing optimizer, LR scheduler, quantization scheme through fairly hacky monkey patching, which can break at any time when updating TT. It would be much nicer & sustainable to have clean entry points like the `ModelHandler` interface I suggest above.

**Comment by tianyu-l** [comment]
@balancap Sounds good to me. Please feel free to make a PR on this, I can help review. Please add relevant tests and docs, following https://github.com/pytorch/torchtitan/blob/main/CONTRIBUTING.md

Besides, in addition to the `ModelHandler`, it'd be good to allow customized optimizer builder and lr scheduler builder, similar to what we are using `parallelize_llama`. I can make a PR on this.

**Comment by tianyu-l** [comment]
@balancap 
It looks that we'll remove the pre/post-optimizer hooks for Float8 in https://github.com/pytorch/torchtitan/pull/812 due to strategy change. So please have a motivating example in your PR / doc.

cc: @mori360

**Comment by tianyu-l** [comment]
For optimizer and lr scheduler, we can follow the idea in #814, or maybe incorporate it directly in it.

For quantization, `ModelHandler` protocol is a proper abstraction, so that one don't need to modify any existing code and only substitute a quantization component, including `train.py`. For bigger changes, we can rely on #814.

Besides, IMO `ModelHandler` protocol is something any quantization component should already follow/define in the first place.

**Comment by balancap** [comment]
@tianyu-l @fegin I opened a draft PR on `ModelHandler` to implement what we discussed: https://github.com/pytorch/torchtitan/pull/823

If you're fine with the design, I can move forward and adding some testing coverage (+ job config integration).

**Comment by balancap** [comment]
@tianyu-l I had not fully appreciated the tensor-parallelism potential issue. For now, in the PR I kept the logic
```python
if enable_float8:
    # TODO(vkuzo): once float8 configuration supports delayed scaling,
    # add a check here to enforce supported float8 all-gather configurations
    # TODO(vkuzo): add the items below to __init__.py of torchao.float8 and import from there
    from torchao.float8.float8_tensor_parallel import (
        Float8ColwiseParallel,
        Float8RowwiseParallel,
        PrepareFloat8ModuleInput,
    )

    rowwise_parallel, colwise_parallel, prepare_module_input = (
        Float8RowwiseParallel,
        Float8ColwiseParallel,
        PrepareFloat8ModuleInput,
    )
else:
    rowwise_parallel, colwise_parallel, prepare_module_input = (
        RowwiseParallel,
        ColwiseParallel,
        PrepareModuleInput,
    )
```
which I guess is fine enough maybe for a first version. 

The alternative could be to pass the `model_handlers` instance to this function and have a function in model handler 
```python
def get tensor_parallel_row_col_prepare(self):
```

**Comment by tianyu-l** [comment]
@balancap 
> which I guess is fine enough maybe for a first version.

Since we only have Float8 right now, I think it looks fine for now, unless you have strong needs to enable other quantization with TP.

From TP perspective, quantization class in torchao should provide an API to return these TP classes, since anyway the customized TP code is there.

If not we can consider defining subclass / subprotocol of ModelHandler e.g. as QuantizationModelHandler, which requires returning customized TP classes.

**Comment by balancap** [comment]
Should close this issue with #823 merged? @vkuzo does it fit what you need, or is there any aspect missing?
